pkg/services/sqlstore/migrations/ualert/channel.go

```diff
@@ -0,0 +1,263 @@
+		}
+
+		settings, secureSettings := migrateSettingsToSecureSettings(m.Type, m.Settings, m.SecureSettings)
+		portedChannels = append(portedChannels, &PostableGrafanaReceiver{
+			Name:                  m.Name,
+			Type:                  m.Type,
+			DisableResolveMessage: m.DisableResolveMessage,
+			Settings:              settings,
+			SecureSettings:        secureSettings,
+		})
+	}
+	receiver.GrafanaManagedReceivers = portedChannels
+
+	n, v := getLabelForRouteMatching(ruleUid)
+	mat, err := labels.NewMatcher(labels.MatchEqual, n, v)
+	if err != nil {
+		return nil, nil, err
+	}
+	route := &Route{
+		Receiver: receiverName,
+		Matchers: Matchers{mat},
+	}
+
+	return receiver, route, nil
+}
+
+// Some settings were migrated from settings to secure settings in between.
+// See https://grafana.com/docs/grafana/latest/installation/upgrading/#ensure-encryption-of-existing-alert-notification-channel-secrets.
+// migrateSettingsToSecureSettings takes care of that.
+func migrateSettingsToSecureSettings(chanType string, settings *simplejson.Json, secureSettings securejsondata.SecureJsonData) (*simplejson.Json, map[string]string) {
+	keys := []string{}
+	switch chanType {
+	case "slack":
+		keys = []string{"url", "token"}
+	case "pagerduty":
+		keys = []string{"integrationKey"}
+	case "webhook":
+		keys = []string{"password"}
+	case "prometheus-alertmanager":
+		keys = []string{"basicAuthPassword"}
+	case "opsgenie":
+		keys = []string{"apiKey"}
+	case "telegram":
+		keys = []string{"bottoken"}
+	case "line":
+		keys = []string{"token"}
+	case "pushover":
+		keys = []string{"apiToken", "userKey"}
+	case "threema":
+		keys = []string{"api_secret"}
+	}
+
+	ss := secureSettings.Decrypt()
+	for _, k := range keys {
+		if v, ok := ss[k]; ok && v != "" {
+			continue
+		}
+
+		sv := settings.Get(k).MustString()
+		if sv != "" {
+			ss[k] = sv
+			settings.Del(k)
+		}
+	}
+
+	return settings, ss
+}
+
+func getMigratedReceiverNameFromRuleUID(ruleUID string) string {
+	return fmt.Sprintf("autogen-panel-recv-%s", ruleUID)
+}
+
+func getLabelForRouteMatching(ruleUID string) (string, string) {
+	return "rule_uid", ruleUID
+}
+
+func extractChannelInfoFromDashboard(d oldDash, panelId int64) (channelUids []string, ruleName string, ruleMessage string, _ error) {
+	panels, err := d.Data.Get("panels").Array()
+	if err != nil {
+		return nil, "", "", err
+	}
+
+	for _, pi := range panels {
+		p := simplejson.NewFromAny(pi)
+		pid, err := p.Get("id").Int64()
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		if pid != panelId {
+			continue
+		}
+
+		a := p.Get("alert")
+
+		ruleMessage = a.Get("message").MustString()
+		ruleName = a.Get("name").MustString()
+
+		// Extracting channel UIDs.
+		uids, err := a.Get("notifications").Array()
+		if err != nil {
+			return nil, "", "", err
+		}
+
+		for _, ui := range uids {
+			u := simplejson.NewFromAny(ui)
+			channelUid, err := u.Get("uid").String()
```

Comment:
There is a possibility that [the notifiers are referenced by the `id` instead of `uid`](https://github.com/grafana/grafana/issues/19771) so I think we need to catch this cases as well and [translate them](https://github.com/grafana/grafana/blob/63b2dd06a5e01587b065cbbe2af8d9a47886b868/pkg/services/sqlstore/alert_notification.go#L167).

Reply:
@papagian I just pushed https://github.com/grafana/grafana/pull/33846/commits/b48e19ab91048c63fd0e6a99afce7189ac6579e4, can you verify if that fixes this? Basically I am including both uid and id in the map now (whichever is available), hence we won't miss the ones that have an ID